File: Backend/EmailClient.Api/Models/DateFilter.cs
namespace EmailClient.Api.Models
{
    /// <summary>
    /// Date filter configuration for filtering emails by date ranges.
    /// Supports both preset periods (days/months/years) and custom date ranges.
    /// </summary>
    public class DateFilter
    {
        /// <summary>
        /// Type of date filter to apply
        /// </summary>
        public DateFilterType FilterType { get; set; } = DateFilterType.All;

        /// <summary>
        /// Number of days to filter (used when FilterType is OlderThanDays)
        /// </summary>
        public int? Days { get; set; }

        /// <summary>
        /// Number of months to filter (used when FilterType is OlderThanMonths)
        /// </summary>
        public int? Months { get; set; }

        /// <summary>
        /// Number of years to filter (used when FilterType is OlderThanYears)
        /// </summary>
        public int? Years { get; set; }

        /// <summary>
        /// Custom start date for filtering (used when FilterType is DateRange)
        /// </summary>
        public DateTime? StartDate { get; set; }

        /// <summary>
        /// Custom end date for filtering (used when FilterType is DateRange)
        /// </summary>
        public DateTime? EndDate { get; set; }

        /// <summary>
        /// Creates a date filter for emails older than specified days
        /// </summary>
        /// <param name="days">Number of days</param>
        /// <returns>Configured DateFilter</returns>
        public static DateFilter OlderThan(int days)
        {
            return new DateFilter
            {
                FilterType = DateFilterType.OlderThanDays,
                Days = days
            };
        }

        /// <summary>
        /// Creates a date filter for emails older than specified months
        /// </summary>
        /// <param name="months">Number of months</param>
        /// <returns>Configured DateFilter</returns>
        public static DateFilter OlderThanMonths(int months)
        {
            return new DateFilter
            {
                FilterType = DateFilterType.OlderThanMonths,
                Months = months
            };
        }

        /// <summary>
        /// Creates a date filter for emails older than specified years
        /// </summary>
        /// <param name="years">Number of years</param>
        /// <returns>Configured DateFilter</returns>
        public static DateFilter OlderThanYears(int years)
        {
            return new DateFilter
            {
                FilterType = DateFilterType.OlderThanYears,
                Years = years
            };
        }

        /// <summary>
        /// Creates a date filter for a custom date range
        /// </summary>
        /// <param name="startDate">Start date (inclusive)</param>
        /// <param name="endDate">End date (inclusive)</param>
        /// <returns>Configured DateFilter</returns>
        public static DateFilter Between(DateTime startDate, DateTime endDate)
        {
            return new DateFilter
            {
                FilterType = DateFilterType.DateRange,
                StartDate = startDate,
                EndDate = endDate
            };
        }

        /// <summary>
        /// Creates a date filter that includes all emails (no filtering)
        /// </summary>
        /// <returns>Configured DateFilter</returns>
        public static DateFilter All()
        {
            return new DateFilter
            {
                FilterType = DateFilterType.All
            };
        }
    }

    /// <summary>
    /// Enumeration of available date filter types
    /// </summary>
    public enum DateFilterType
    {
        /// <summary>
        /// No date filtering - include all emails
        /// </summary>
        All,

        /// <summary>
        /// Filter emails older than specified number of days
        /// </summary>
        OlderThanDays,

        /// <summary>
        /// Filter emails older than specified number of months
        /// </summary>
        OlderThanMonths,

        /// <summary>
        /// Filter emails older than specified number of years
        /// </summary>
        OlderThanYears,

        /// <summary>
        /// Filter emails within a custom date range
        /// </summary>
        DateRange
    }
}

File: Frontend/src/types/DateFilter.ts
/**
 * Enumeration of supported date filter types for email filtering.
 * Allows users to filter emails by preset periods or custom date ranges.
 */
export enum DateFilterType {
  /** Show all emails without date filtering */
  All = "All",
  /** Filter emails older than specified number of days */
  OlderThanDays = "OlderThanDays",
  /** Filter emails older than specified number of months */
  OlderThanMonths = "OlderThanMonths",
  /** Filter emails older than specified number of years */
  OlderThanYears = "OlderThanYears",
  /** Filter emails within a custom date range */
  DateRange = "DateRange",
}

/**
 * Date filter configuration for filtering emails by date criteria.
 * Supports both preset periods and custom date ranges for flexible email management.
 * Property names match the backend C# model exactly.
 */
export interface DateFilter {
  /** Type of date filter to apply */
  FilterType: DateFilterType;

  /** Number of days for OlderThanDays filter type */
  Days?: number;

  /** Number of months for OlderThanMonths filter type */
  Months?: number;

  /** Number of years for OlderThanYears filter type */
  Years?: number;

  /** Start date for DateRange filter type */
  StartDate?: Date;

  /** End date for DateRange filter type */
  EndDate?: Date;
}

/**
 * Request model for retrieving emails with optional date filtering.
 * Used to send filter criteria to the backend API.
 * Property names match the backend C# model exactly.
 */
export interface EmailsByFilterRequest {
  /** Optional date filter to apply when retrieving emails */
  DateFilter?: DateFilter | null;
}

/**
 * Utility functions for creating common date filters
 */
export class DateFilterHelpers {
  /**
   * Creates a date filter for emails older than specified days
   */
  static olderThan(days: number): DateFilter {
    return {
      FilterType: DateFilterType.OlderThanDays,
      Days: days,
    };
  }

  /**
   * Creates a date filter for emails older than specified months
   */
  static olderThanMonths(months: number): DateFilter {
    return {
      FilterType: DateFilterType.OlderThanMonths,
      Months: months,
    };
  }

  /**
   * Creates a date filter for emails older than specified years
   */
  static olderThanYears(years: number): DateFilter {
    return {
      FilterType: DateFilterType.OlderThanYears,
      Years: years,
    };
  }

  /**
   * Creates a date filter for a custom date range
   */
  static between(startDate: Date, endDate: Date): DateFilter {
    return {
      FilterType: DateFilterType.DateRange,
      StartDate: startDate,
      EndDate: endDate,
    };
  }

  /**
   * Creates a date filter that includes all emails (no filtering)
   */
  static all(): DateFilter {
    return {
      FilterType: DateFilterType.All,
    };
  }

  /**
   * Common preset filters for quick selection
   */
  static readonly PRESETS = {
    ALL: DateFilterHelpers.all(),
    LAST_7_DAYS: DateFilterHelpers.olderThan(7),
    LAST_30_DAYS: DateFilterHelpers.olderThan(30),
    LAST_90_DAYS: DateFilterHelpers.olderThan(90),
    LAST_6_MONTHS: DateFilterHelpers.olderThanMonths(6),
    LAST_YEAR: DateFilterHelpers.olderThanYears(1),
    LAST_2_YEARS: DateFilterHelpers.olderThanYears(2),
  };
}
